### ISSUE TITLE
Adding new characteristics to the HLG visualizations

### ISSUE THREAD

**Issue by freyam** [dask/dask#7847]
- [x] Task 1 -  #7869 
**Status**: Merged

- [x] Task 2A - #7886
**Status**: Review

~~- [ ] Task 2B - (SVG implementation in Graphviz) #XXXX~~
~~**Status**: Graphviz acting weird and not allowing us to downscale the images~~

(I will keep updating this comment)

----

Hi :wave:

I am posting this Feature Request issue here to start a discussion regarding enhancing the Graphviz output of the `dask.visualize()` method.

While I was adding color to show whether a HLG layeri s materialized or not using a light gray fill color, @mrocklin pointed out that it would be better to make this discussion public to gather opinions from the Dask community as a whole.

> In the future I think that users would be very interested in attributes
> like layer size and type, as well as collection attributes like chunking
> structure in the case of dask arrays. Personally I would encourage you to
> focus your efforts there. You might also want to raise issues with
> proposed changes in the future. That will help you to get feedback on idea
> from a broad set of people in the community, rather than just the one or
> two that are engaging in the gsoc slack channel.

Source: PR #7843

@GenevieveBuckley worked on #7309 and added a new dictionary `collection_annotations` which has crucial information about the High Level Graphs which I believe can be shown in some way on the Graphviz output. 

@martindurant mentioned this over at https://github.com/dask/dask/issues/7301#issuecomment-860686336

> ...  based on the task naming conventions and e.g., for high-level graphs the number of sub-tasks and for arrays, the size of the operands. Such information might be added into the nodes as text, colour or edge-style (all probably optional).

If anyone else has any ideas, please leave them in the comments section. How should I proceed?

Let's make the output of dask.visualize() more interesting and appealing to the eye! :raised_hands:

**Comment by GenevieveBuckley** [comment]
I'd echo that we:
1. shape (eg: the whole array shape, the number of columns in a dataframe, maybe dataframe length?)
2. chunks / partition shapes
2. the type of dask object (eg: dask Array, Dataframe, etc)
3. the type of chunks/partitions (eg: it's an array with numpy chunks, a dataframe with pandas partitions)

#1 and #2 are probably the most interesting things in this list. I know we've already talked about this a lot over on Slack, feel free to ask questions.

Here's a tiny example that shows the shape & chunksize changing:
```python
import dask.array as da
x = da.arange(1000).reshape(1, 1000)
y = x.T
z = x * y
result = da.sum(z)
```

Note: we can add more things to the `collection_annotations` dictionary if you need that for visualization. That's flexible.

**Comment by mrocklin** [comment]
I think that it is useful to break up attributes that can be applied to all layers and those that can be applied to layers coming from a specific collection type, like dataframe or array.  I'll list a few of these below, copying some from @GenevieveBuckley 's list above

### General attributes applied to all layers

-  Number of tasks `len(layer)`.  I think it seems natural to associate this with size.  The more tasks in a layer the "heavier" it should feel
- The collection type of the layer, although hopefully this becomes obsolete once we implement collection-type-specific attributes

Yeah, so now that I think about it, there isn't much that is generally available.  Number of tasks would be cool though and is probably a good place to start.  I think that it is a useful attribute to convey.  It draws the eye to important parts of the graph.  It should also be easy to compute in general.

### Specific attributes for arrays

As we've seen in Tom White's video, the chunk structure is big.  We have code to generate chunking images as SVG.  I believe that DOT can consume SVG.  Putting chunk structure and dtype information into every node in the high level graph would, I think, solve Tom's problem.  I think that this would be a **big win** and we already have most of the tricky code to do it.  Someone would need to learn more about how to embed SVG into DOT in order to make this happen.

We can also override the size <-> number of tasks relationship with the amount of memory the layer should take up.  Now that we have size and dtype we should be able to make a more accurate assesment of size.  

### Specific attributes for dataframes

Nothing immediately comes to mind here for me.  We probably don't want to list all of the columns or anything (that can be hundreds of names in practice, and seems hard to summarize).  A simple thing we can do would be to include the chunk type, like a Pandas or RAPIDS logo off in the corner of the box.

**Comment by mrocklin** [comment]
So, if I were to do this I would probably start with either ...

1.  Size, as computed by `len(layer)`.  Given that sizes can vary by several orders of magnitude (1 to 100000) we will have to think of some function to transform this dynamic range (sigmoid, log, ...)
2. Putting SVG images into boxes.  This would be directly useful for array chunks, and would also be a good tool to have.  Once we know how to use SVG then we have a very powerful tool in our toolbox for future work.

**Comment by GenevieveBuckley** [comment]
>So, if I were to do this I would probably start with either ...
>
> 1. Size, as computed by `len(layer)`.  Given that sizes can vary by several orders of magnitude (1 to 100000) we will have to think of some function to transform this dynamic range (sigmoid, log, ...)

+1 to trying log scaling first for the box sizes. I would not spend lots of time trying lots of different options (that can come much later on if need be, like after the first GSOC evaluation)

> 2. Putting SVG images into boxes.  This would be directly useful for array chunks, and would also be a good tool to have.  Once we know how to use SVG then we have a very powerful tool in our toolbox for future work.

I might suggest a two-part approach
2a: Add the array SVG images to the HTML high level graph layer representation (an easy, good first step, that will get us a feature that's been requested, and get you some familiarity with the SVG part of the code)
2b: Add the array SVG images into the graphviz boxes..

**Comment by GenevieveBuckley** [comment]
More detail for 2a, above.
Adding array SVG images to the HTML representation was requested here https://github.com/dask/dask/pull/7763#issuecomment-858589993
> In some more aggressive future we might consider including the dask array SVG array chart directly in the HTML, that's probably future work though

Suggested approach:
* Add `self.chunks` into the `collection_annotations` dictionary. That part of the code is here: https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/array/core.py#L1159
* Use the `svg` function to draw the array picture https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/array/svg.py#L7
* Add the array picture into the `_repr_html_` method on the `Layer` class https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/highlevelgraph.py#L474
* Tidy up the info we see in the HTML table. For large arrays with many chunks, there wasn't a lot of use in reporting the fully specified chunks - that turns out to be really long, you can see this in Julia's comment [here](https://github.com/dask/dask/pull/7763#issuecomment-858658497). So, I suggest when we make the HTML table, first we make a copy of the `collection_annotations` dictionary ([like this](https://www.programiz.com/python-programming/methods/dictionary/copy)), then we pop the chunks key out, and finally we make the HTML table from the copied & altered dictionary. That part of the code is here:
 https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/highlevelgraph.py#L528

**Comment by freyam** [comment]
Yes, I will get started on this🙌

**Comment by freyam** [comment]
# Idea for showing size characteristics (Task `#1`)
(discussed with @GenevieveBuckley)

Every layer has `X` number of tasks associated with it. This can range from 1 - 100,000. One thing we can do is create a static scale representation for the proposed number of tasks and change the font size (which in turn changes the node size) of individual layers. 

## Demo

```py
import dask.array as da

a = da.ones((20), chunks=(2))

b = da.dot(a, a) + da.dot(a, a) + da.dot(a, a)
c = da.dot(a, a) + da.dot(a, a) + da.dot(a, a)

c = c + 10

c = c + da.dot(a, a)

c = c - 5

c.dask.visualize()
```

### Current Graphviz Graph 

![image](https://user-images.githubusercontent.com/62539811/124229659-01775900-db1f-11eb-9b06-3bdfcd1ddbf7.png)


### Variant 1 (Adding a new label stating the number of tasks associated with each layer node)

![image](https://user-images.githubusercontent.com/62539811/124229757-279cf900-db1f-11eb-87f8-721ce87cd68b.png)

### Variant 2 (Variant  1 + Dynamic Node Size depending on the number of tasks)

**Larger the node, the more tasks**

![image](https://user-images.githubusercontent.com/62539811/124229829-3e435000-db1f-11eb-9fc1-00ebda9632f0.png)

### Variant 3 :100: (Variant 1 + Dynamic Node Color depending on the number of tasks)

**Darker the gray shade, the more tasks**

![image](https://user-images.githubusercontent.com/62539811/124230409-038de780-db20-11eb-9752-c5149e8b8738.png)

### Variant 4 (Variant 2 + Variant 3)

![image](https://user-images.githubusercontent.com/62539811/124231509-692ea380-db21-11eb-91e7-0546c54f2f07.png)


## Plan
I personally like Variant 3. It's simple and colorful (monochromatic :sparkles:).
The specifics of size/color can be discussed. My plan was to use a predefined scale for both.

### Size
- Minimum Node Size = `10`
- Maximum Node Size = `50`
The intermediate node sizes can be scaled from a log function applied to the `n_tasks`

### Color (gray)
- Lightest Shade = `gray99` (#FCFCFC)
- Darkest Shade  = `gray78` (#C7C7C7)
The intermediate font sizes can be scaled from a log function applied to the `n_tasks`

One problem with https://github.com/dask/dask/pull/7843 was that there was no indication of what the shading meant. But, here, we also provide the information about the number of tasks for each layer before coloring it based on that itself.

![image](https://user-images.githubusercontent.com/62539811/124231086-e279c680-db20-11eb-9005-fe3b05523392.png)
(ignore `gray8` and `gray9`)


Does this sound good? @mrocklin 
I can implement these into the `dask.visualize()` with an extra attribute (can be named anything).

Example: `dask.visualize(fancy=True)` or `dask.visualize(chracteristics="Variant1"); dask.visualize(chracteristics="Variant2"); dask.visualize(chracteristics="Variant3");` and so on

**Comment by mrocklin** [comment]
First, thank you for putting up a few choices to talk through.  This is
really helpful to engage conversation, and seems much more time efficient
than jumping directly into code.

I expect that we will end up using shading for other attributes.  I don't
know though.  If so, then size seems like the most natural thing to
associate to the number of nodes.

Also, I'm not sure yet that we will want to include the number of tasks in
a table.  Instead, I expect that area to be layer-specific.  I would start
with adding just size and see how that goes.  (maybe this was your
intention anyway, and including the number of nodes was just for
explanation here, in which case please ignore my comment).

**Comment by martindurant** [comment]
I think the sizing is a nice idea, but I would suggest a reasonable minimum and maximum size. You do think there is scope fr colour here, though, especially if we choose to surface various options to the user in `.visualize()` and we currently do with `color="order"`. To repeat myself, there are many attributes of each node that can be changed (font size, node size, shape, border colour/thickness, fill ...), as well as what text to include, so I think that getting the attributes into the plotting code is the primary thing, and deciding the how to represent them secondary. I defer to deciding on what information is important about to each layer type. My personal favourite would be some manner of "type": IO | CPU | GPU | shuffle ...
Nevertheless, examples like the one above really help understand the implications!

I agree that getting HTML/SVG into the nodes would be a massive improvement and allow for tying together the various visual elements. If layer types can all represent themselves, then we can dispatch to them and accept that some are plainer than others.

By the way, you might want to get a tool like colorcet to generate suitable colour schemes rather than trying to pick HTML names. colorcet can generate a set of "#xxxxxx"-style values.

**Comment by martindurant** [comment]
by the way, shouldn't it have been:
- N=1, fontsize=10
- N=10, fontsize=10
- N=50, fontsize=17
- N=100,fontsize=20
- N=750, fontsize=28
- N=1000, fontsize=30

**Comment by freyam** [comment]
> Also, I'm not sure yet that we will want to include the number of tasks in a table. Instead, I expect that area to be layer-specific. I would start with adding just size and see how that goes. (maybe this was your intention anyway, and including the number of nodes was just for explanation here, in which case please ignore my comment).

@mrocklin In reality, I was referring to the number of tasks that each layer (a node in an HLG) contains.

n_tasks can be calculated using `len(layer)`
![image](https://user-images.githubusercontent.com/62539811/124327264-d6315000-db98-11eb-912c-3be0ecce72e4.png)

Did you mean the chunk size when you mentioned "size"? or the number of tasks? After discussing with @GenevieveBuckley, I decided to start with the number of tasks, and then experiment with chunk size/shape depending on whether it's a Dask array/dataframe.

or were you talking about the string "n_tasks" in the graph output? it does look redundant. How else would you like to represent that on the graph?

There are several other attributes (https://graphviz.org/doc/info/attrs.html) which can be used instead of just fontsize.

**Comment by freyam** [comment]
> I defer to deciding on what information is important about to each layer type. My personal favourite would be some manner of "type": IO | CPU | GPU | shuffle ...

This seems like a neat idea. Can you expand upon it?
@martindurant

**Comment by GenevieveBuckley** [comment]
(FYI @freyam, Matt's on vacation this week)

**Comment by freyam** [comment]
Oh, I see. I will start the PR then.

**Comment by GenevieveBuckley** [comment]
> > I defer to deciding on what information is important about to each layer type. My personal favourite would be some manner of "type": IO | CPU | GPU | shuffle ...
> 
> This seems like a neat idea. Can you expand upon it?
> @martindurant

To expand on this a little:
* You can probably make a list of IO layer classes (eg: `DataFrameIOLayer`, etc.) to check against.
* You can get some sense of whether dask dataframe or dask array computations are happening on the GPU by examining the type of dask chunks/partitions. This information is currently recorded in the `collection_annotations` dictionary attribute for high level graph layers for dataframes & arrays (we won't worry too much about bags or non-array, non-dataframe things for this at the moment).
    * Dask dataframes will use [cudf](https://github.com/rapidsai/cudf) partitions (instead of pandas) for GPU computation
    * Dask arrays will use [cupy](https://github.com/cupy/cupy) chunks (instead of numpy) for GPU computation

**Comment by freyam** [comment]
I have a question:
What is the general range of an HLG layer's task count?

I would like to know the following:
- a reasonable minimum value for `n_tasks`
- a reasonable maximum value for `n_tasks` 
- a reasonable math function to account for all kinds of values for `n_tasks` (I am inclining towards [`numpy.log`](https://numpy.org/doc/stable/reference/generated/numpy.log.html) and [`numpy.clip`](https://numpy.org/doc/stable/reference/generated/numpy.clip.html))

Alternate Idea: I can calculate the minimum and the maximum `n_tasks` and use them instead. 
Pro: This would normalize the overall graph structure. The graphs wouldn't look very disproportionate.
Con: We can no longer compare two different task graphs on the basis of the size as the minimums and the maximums can differ. (CounterPoint to the Con: We are actually also mentioning the actual number of `n_tasks` on every node itself. So, maybe this wouldn't cause much trouble) 

-----

@GenevieveBuckley This seems like a great idea. It also seems fine to implement. All that's left is how do we want to represent it on the screen? I believe the implementation of the "adding to the graphviz" is quite trivial. The really tricky part is how we want to show it. Which traits of the node/edge would you wanna tweak to show the new information. I do have some amazing sample diagrams ready which I can draft up by tomorrow. But, if anyone here has any plans or suggestions,  I would love to hear that. 

> * You can get some sense of whether dask dataframe or dask array computations are happening on the GPU by examining the type of dask chunks/partitions. This information is currently recorded in the `collection_annotations` dictionary attribute for high level graph layers for dataframes & arrays (we won't worry too much about bags or non-array, non-dataframe things for this at the moment).
>   
>   * Dask dataframes will use [cudf](https://github.com/rapidsai/cudf) partitions (instead of pandas) for GPU computation
>   * Dask arrays will use [cupy](https://github.com/cupy/cupy) chunks (instead of numpy) for GPU computation


In short, we already have all the required information stored in the `hg` HLG object or the `collection_annotations` dict. All we need to know right now is how we wish to show the new information to the users. Graphviz is pretty cool and supports so many attributed and it's all up to us to use them. 
Reference; http://www.graphviz.org/doc/info/attrs.html

For `Task #1`, I have created a PR #7869

**Comment by GenevieveBuckley** [comment]
> I have a question:
> What is the general range of an HLG layer's task count?

This could vary a lot. The dataframe shuffle example is probably the smallest size reasonable example we have (instead of the tiny toy examples we've also looked at). But depending on what users are doing, it could be very, very large indeed. I don't think we can choose a fixed upper value for this.

> I would like to know the following:
> 
> * a reasonable minimum value for `n_tasks`

I'd say 1 task is the minimum value possible.

> * a reasonable maximum value for `n_tasks`

I don't think there is a single maximum value we can pick. It will vary wildly depending on what kind of computation the user is doing. (Your suggestion to normalize to the biggest layer in each HLG structure might be a good way to handle this)

> * a reasonable math function to account for all kinds of values for `n_tasks` (I am inclining towards [`numpy.log`](https://numpy.org/doc/stable/reference/generated/numpy.log.html) and [`numpy.clip`](https://numpy.org/doc/stable/reference/generated/numpy.clip.html))

A log scale is a good option, yes. 

> Alternate Idea: I can calculate the minimum and the maximum `n_tasks` for each HLG and use the local minima and local maxima instead.
> Pro: This would normalize the overall graph structure. The graphs wouldn't look very disproportionate.
> Con: We can no longer compare two different task graphs on the basis of the size as the minimums and the maximums can differ. (CounterPoint to the Con: We are actually also mentioning the actual number of `n_tasks` on every node itself. So, maybe this wouldn't cause much trouble)

I think this is a good idea. 

> @GenevieveBuckley This seems like a great idea. It also seems fine to implement. All that's left is how do we want to represent it on the screen? I believe the implementation of the "adding to the graphviz" is quite trivial. The really tricky part is how we want to show it. Which traits of the node/edge would you wanna tweak to show the new information. I do have some amazing sample diagrams ready which I can draft up by tomorrow. But, if anyone here has any plans or suggestions, I would love to hear that.

I had assumed Martin's suggestion was to show a different color for each of these categories. Scrolling back up, it looks like he didn't actually say that & I just imagined it. Nevertheless, perhaps color is good place to start. (Also, as I said early on in this project, you will probably have ideas about the best way to represent certain characteristics visually. Definitely add your own suggestions or ideas for discussion here too, if you have them)

I also agree with Martin's comment *"I think that getting the attributes into the plotting code is the primary thing, and deciding the how to represent them secondary"*

**Comment by GenevieveBuckley** [comment]
> ### Specific attributes for arrays
> 
> As we've seen in Tom White's video, the chunk structure is big. We have code to generate chunking images as SVG. I believe that DOT can consume SVG. Putting chunk structure and dtype information into every node in the high level graph would, I think, solve Tom's problem. I think that this would be a **big win** and we already have most of the tricky code to do it. Someone would need to learn more about how to embed SVG into DOT in order to make this happen.
> 
> We can also override the size <-> number of tasks relationship with the amount of memory the layer should take up. Now that we have size and dtype we should be able to make a more accurate assesment of size.

I'm unsure if I've linked to this video elsewhere, but since I don't see a link in *this* thread I'm adding one for convenience.

@freyam, this is [Tom White's video](https://www.youtube.com/watch?v=qt6YsHoPpZs&list=PLJ0vO2F_f6OBAY6hjRHM_mIQ9yh32mWr0&index=2) and [this specific section](https://youtu.be/qt6YsHoPpZs?list=PLJ0vO2F_f6OBAY6hjRHM_mIQ9yh32mWr0&t=303) is most relevant to our conversation, where he talks about how it's difficult to examine the shape of all the intermediate arrays in a computation. (This is what we're trying to solve with tasks 2a and 2b, discussed above)

**Comment by tomwhite** [comment]
Thanks @GenevieveBuckley and @freyam for your work on this.

In case it's useful, here's a bit more information on the case I was working on. What I did at the time was create a notebook to break down each operation so I could see the chunk structure of each intermediate array: https://nbviewer.jupyter.org/github/tomwhite/gwas-benchmark/blob/1a73ff865d100da7ea51c73456aa1a8526ff29b7/gwas_simulation_shapes.ipynb.

Simplifying further, what I'd like to see is the shape and chunk structure for the following arrays, ideally by inspecting the HLG for `XLP`:
```python
import dask.array as da
XL = da.ones((67552, 365941), chunks=(652, 5792))
XL2 = XL.T.astype("f4")
XC = da.ones((365941, 26), chunks=(365941, 26))
LS = da.linalg.lstsq(XC, XL2)[0]
XC_LS = XC @ LS
XLP = XL2 - XC_LS
```

**Comment by freyam** [comment]
I have implemented Variant 2 (as shown above) over at #7869 (`Task #1`).

**Comment by freyam** [comment]
### - [x] Task `#1` (Size Characteristics in Graphviz) #7869 
**Status**: Waiting for Merge

### - [x] Task `#2A` (SVG implementation in HTML Repr) #7886
**Status**: Pending Review

### - [ ] Task `#2B` (SVG implementation in Graphviz) #XXXX
**Status**: Graphviz acting weird and not allowing us to downscale the images

(I will keep updating this comment)

**Comment by freyam** [comment]
@tomwhite 

## Demo

```py
import dask.array as da

XL = da.ones((67552, 365941), chunks=(652, 5792))
XL2 = XL.T.astype("f4")
XC = da.ones((365941, 26), chunks=(365941, 26))
LS = da.linalg.lstsq(XC, XL2)[0]
XC_LS = XC @ LS
XLP = XL2 - XC_LS

XLP.dask
```

![image](https://user-images.githubusercontent.com/62539811/126084254-c2fb02c3-0c64-4336-8ba8-8ad4d204fff8.png)
![image](https://user-images.githubusercontent.com/62539811/126084264-864ab814-dcfc-47bb-8a59-46fdf3dcd208.png)
![image](https://user-images.githubusercontent.com/62539811/126084267-a9f67f6c-93b6-4ee8-ae2c-b8ff0d625b2f.png)
![image](https://user-images.githubusercontent.com/62539811/126084269-472bbc95-8b39-405c-848b-d317bc41e39d.png)
![image](https://user-images.githubusercontent.com/62539811/126084276-af75f8a0-6e22-4943-a90a-6c47641e56a9.png)
![image](https://user-images.githubusercontent.com/62539811/126084281-372a6f5e-e59f-4c35-a0d7-deaaa5f7611d.png)

The computation looks incredible! :fire:

**Comment by freyam** [comment]
## Updates

1. **Task 1** has been added #7869 :tada: 
2. **Task 2A** is complete and waiting for approval #7886 :100: 
3. **Task 2B** doesn't seem like a good fit in the graphical representation as it takes a lot of space and it seems more practical for users to check the HTML repr for the array SVGs at every stage of the computation (Task 2A). Plus, Graphviz is not very great at showing images inside nodes. So, it's a bottleneck right now and has been abandoned.

**Comment by freyam** [comment]
@tomwhite 

#7886 has been merged!  :tada: 

Now, you will be able to visualize the intermediate state of the Dask arrays within the HTML Representation and it looks amazing :smile:

**Comment by tomwhite** [comment]
> Now, you will be able to visualize the intermediate state of the Dask arrays within the HTML Representation and it looks amazing 😄

That's great - thanks for working on this @freyam!

**Comment by GenevieveBuckley** [comment]
Would love to hear your experience @tomwhite if you road test at any point. We're hoping it should make for an easier debugging experience with some of the common frustrations you've shared with us.
